### PR TITLE
introduce per-file timing-stats

### PR DESCRIPTION
**Pull request by huguesb**
When profiling mypy over a large codebase, it can be useful
to know which files are slowest to typecheck.

Gather per-file timing stats and expose them through a new
(hidden) command line switch

### CODE REVIEW

File: mypy/build.py [hunk 1]
```diff
@@ -1808,6 +1810,9 @@ class State:
 
     fine_grained_deps_loaded = False
 
+    # Cumulative time spent on this file (for profiling stats)
```
Comment (JukkaL) [inline mypy/build.py:1813]: Include the unit of time in the name of the variable (e.g. `time_spent_ns`).

File: mypy/build.py [hunk 2]
```diff
@@ -3091,6 +3120,8 @@ def process_graph(graph: Graph, manager: BuildManager) -> None:
         manager.log("No fresh SCCs left in queue")
 
 
+
+
```
Comment (JukkaL) [inline mypy/build.py:3124]: These extra empty lines seem redundant.

**Review by JukkaL — COMMENTED** [review]
Cool, I've wanted to have this feature myself when debugging performance issues.

Can you add at least one integration test that ensures that mypy can be run with the flag successfully? Possibly also check the format of the generated file. Obviously the timings aren't repeatable, so they would have to be ignored.

Also, the builds were failing.

**Comment by huguesb** [comment]
> Can you add at least one integration test that ensures that mypy can be run with the flag successfully? Possibly also check the format of the generated file. Obviously the timings aren't repeatable, so they would have to be ignored. 
Done. I added a way to specify test output files via regex to achieve that.

> Also, the builds were failing.
Yeah, I didn't realize mypy still supported py3.6
Fixed

File: mypy/build.py [hunk 3]
```diff
@@ -2771,6 +2790,16 @@ def dumps(self) -> str:
                                                      json.dumps(self.deps))
 
 
+def dump_timing_stats(path: str, graph: Graph) -> None:
+    """
+    Dump timing stats for each file in the given graph
+    """
+    with open(path, 'w') as f:
+        for k in sorted(graph.keys()):
```
Comment (JukkaL) [inline mypy/build.py:2798]: What about sorting by the time spent in file instead of the module name? It will be easy to search by module name within the stats dump, but sorting by the time is slightly more involved.

**Review by JukkaL — COMMENTED** [review]
Looks good, just a few optional comments.

File: mypy/build.py [hunk 4]
```diff
@@ -2188,8 +2201,10 @@ def compute_dependencies(self) -> None:
     def type_check_first_pass(self) -> None:
         if self.options.semantic_analysis_only:
             return
+        t0 = time.perf_counter()
         with self.wrap_context():
             self.type_checker().check_first_pass()
+        self.time_spent_us += int((time.perf_counter() - t0) * 1e6)
```
[the remark this answers is not in the document]
Comment (huguesb) [reply mypy/build.py:2207]: sure, that seems fine. That can then be refactored  to use `perf_counter_ns` whenever support for py3.6 is dropped

File: mypy/build.py (hunk 3 above)
Comment (huguesb) [reply mypy/build.py:2798]: A straightforward `sort -n -k2` on the output can give you time-sorting from the current  format.

I originally thought about the module order being nice to get diffs between runs but the jitter is going to be high enough that such a diff would probably need a little extra massaging (e.g only show changes above a certain % threshold) to actually be helpful.

TL;DR I don't think the sort order makes much of a difference in practice so I would be fine going with either order, or even letting it be unsorted. In follow-up commit we can introduce a script similar to go's `benchstat` to get more value out of the raw data.

**Review by JukkaL — APPROVED** [review]
Thanks! Looks good.

**Comment by 97littleleaf11** [comment]
cool feature!

**Comment by huguesb** [comment]
On Thu, Apr 21, 2022, 7:54 AM Jukka Lehtosalo ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Cool, I've wanted to have this feature myself when debugging performance
> issues.
>
> Can you add at least one integration test that ensures that mypy can be
> run with the flag successfully? Possibly also check the format of the
> generated file. Obviously the timings aren't repeatable, so they would have
> to be ignored.
>
will do

> Also, the builds were failing.
>
yeah,I didn't realize mypy still supported running under 3.6. I will need
to use a different timing API

**Comment by JelleZijlstra** [comment]
@huguesb did you comment on the wrong PR? For what it's worth we dropped 3.6 support.

**Comment by huguesb** [comment]
uh, that's very weird. I did not intentionally send that... it looks relevant to the history of this PR but not after its closure, and I don't even see this message as a "sent" item my email client...

**Comment by JelleZijlstra** [comment]
Maybe some email queue in GitHub had a very long delay :)
